csp_billing_adapter/archive.py

```diff
@@ -44,17 +49,34 @@ def append_metering_records(
         metering and usage records to be archived.
     :param max_length:
         The max size of the archive list.
+    :param max_bytes:
+        The max size in bytes of the archive.
     :return:
         The archive of meterings and usage records with the
         billing_record appended. If archive ends up greater
-        than max lengh the first (oldest) record is dropped.
+        than max lengh or max bytes the archive is trimmed
+        as necessary to satisfy both max_length and
+        max_bytes.
     """
     archive.append(billing_record)
 
     if len(archive) > max_length:
-        return archive[1:]
-    else:
-        return archive
+        archive = archive[1:]
+
+    if max_bytes > 1:
+        # Treat 0 and 1 the same. Disable max bytes option.
+        # This prevents infitite loop when value is 1 since
+        # empty list is 2 bytes.
+        while True:
+            # Trim archive until it is smaller than max bytes
+            archive_size = len(bytes(json.dumps(archive), 'utf-8'))
```

Comment:
The calculation is not correct, the `bytes()` function turns the string that `json.dumps()` produces into a list of integers, an int for every character. Then `len()` gives us the number of integers in the list. However, this is not equal to the size of the data structure from a memory footprint perspective. An integer object in Python consumes 28 bytes
```
>>> import sys
>>> sys.getsizeof(1)
28
```
Meaning the calculated `archive_size` needs to be multiplied by `28` for the comparison to `max_bytes` since we are concerned to not exceed the memory limit of a config map in k8s.

Reply:
No it's correct:

```
import json


archive = [
    {
        "bill_time": "2024-01-03T20:06:42.076972+00:00",
        "metering_id": "FFFFFFFF-FFFF-FFFF-FFFF-FFFFFFFFFFFF",
        "usage_records": [
            {
                "managed_node_count": 9,
                "managed_systems": 6,
                "reporting_time": "2024-01-03T20:06:42.076972+00:00"
            }
        ]
    }
]

archive_size = len(bytes(json.dumps(archive), 'utf-8'))

with open('/tmp/t.json', 'w') as f:
    json.dump(archive, f)

print(f'computed size: {archive_size}')
```

```
16:37:02 ▶ python3 /tmp/t.py
computed size: 228
16:37:05 ▶ l /tmp/t.json
-rw-r--r-- 1 smarlow smarlow 228 Jan 10 16:37 /tmp/t.json
```

Reply:
I guess there's no need to convert to bytes though. It should only contain ascii characters which are 1 byte per character.

Reply:
getsizeof is returning the size of the python object not the size of the character that will be stored.